<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Avt by HFSVBC</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>Avt</h1>
        <h2>Allele Validation Tool (AVT) for whole genome Multilocus Sequence Typing (wgMLST) schemas</h2>
        <a href="https://github.com/HFSVBC/AVT" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <p>Hugo Curado(1), Margarida Cândido(1), Mickael Silva(2), João André Carriço(2) e Francisco M. Couto(1)</p>

<p>(1) LaSIGE, Departamento de Informática, Faculdade de Ciências, Universidade de Lisboa, Lisboa, Portugal<br>
(2) Instituto de Microbiologia, Instituto de Medicina Molecular, Faculdade de Medicina, Universidade de Lisboa, Lisboa, Portugal</p>

<h1>
<a id="abstract" class="anchor" href="#abstract" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Abstract</h1>

<p>With High Throughput Sequencing methods (HTS), sequence-based microbial
identification methods have evolved to query thousands of loci present in a given
bacterial species. Methods such as the ones used in BigsDB(1), are based on draft
genomes (DGs) produced by de novo assemblies from reads, and BLAST based
methodologies match the loci on DGs to existing allelic databases. However when novel
alleles are found, they should be validated since the variation could be an artifact
introduced by the de novo assembler. Thus, the role of our work is to create AVT
(Allele Validation Tool), a tool that is capable of validate novel alleles found in de
novo assemblers by allele calling algorithms.The proper validation of an allele can be of
extreme importance, specially in the correct detection of outbreak situations based on
wgMLST.
The inputs for AVT are the reads of the strains under study, the contigs
produced by the de novo assembler and a list of all the novel alleles found. AVT uses a
sequence mapper such as Bowtie2 (2), and samtools/bfctools (3), to detect any SNPs
and validate coverage over the length of the allele. The output will report any
SNP/indels found or if the allele is confirmed.
As future work, we plan on providing the tool as a web application and also as a
standalone application (based on python and browser compatible languages) so
everyone can access it easily. Moreover, we also want to make our code open source so
everyone can contribute and improve it to fit their best needs.</p>

<h2>
<a id="references" class="anchor" href="#references" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>References</h2>

<ul>
<li>(1) Jolley, KA and Maiden M. BMC Bioinformatics 2010 11:595.</li>
<li>(2) Langmead B, Salzberg S Nature Methods. 2012, 9:357-359.</li>
<li>(3) Li H. Bioinformatics. 2011 Nov 1;27(21):2987-93.</li>
</ul>

 <h2>
<a id="ccknowledgements" class="anchor" href="#acknowledgements" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Acknowledgements</h2>
<p>Hugo Curado and Margarida Cândido have worked equally on writting this article.</p>

        </section>

        <aside id="sidebar">
          <a href="https://github.com/HFSVBC/AVT/archive/master.zip" class="button"><small>Download</small>ZIP</a>
          <p class="repo-owner"><a href="https://github.com/HFSVBC/AVT"></a> is maintained by <a href="https://github.com/HFSVBC">HFSVBC</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

  
  </body>
</html>
